src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -446,6 +396,158 @@ namespace AppInstaller::Repository
+                        // Name not already defined, add it
+                        m_sourceList.emplace_back(std::move(source));
+                    }
+                    else
+                    {
+                        auto& found = GetSource(source.Name);
+                        AICLI_LOG(Repo, Info, << "Source named '" << found.Name << "' is already defined at origin " << ToString(found.Origin) <<
+                            ". The source from origin " << ToString(origin) << " is dropped.");
+                    }
+                }
+            }
+
+            auto metadata = GetMetadata();
+            for (const auto& metaSource : metadata)
+            {
+                if (ContainsSource(metaSource.Name))
+                {
+                    GetSource(metaSource.Name).LastUpdateTime = metaSource.LastUpdateTime;
+                }
+            }
+        }
+
+        std::vector<std::reference_wrapper<SourceDetailsInternal>> SourceListInternal::GetCurrentSourceRefs()
+        {
+            std::vector<std::reference_wrapper<SourceDetailsInternal>> result;
+
+            for (auto& s : m_sourceList)
+            {
+                if (!s.IsTombstone)
+                {
+                    result.emplace_back(std::ref(s));
+                }
+                else
+                {
+                    AICLI_LOG(Repo, Info, << "GetCurrentSourceRefs: Source named '" << s.Name << "' from origin " << ToString(s.Origin) << " is a tombstone and is dropped.");
+                }
+            }
+
+            return result;
+        }
+
+        auto SourceListInternal::FindSource(std::string_view name, bool includeTombstone)
+        {
+            return std::find_if(m_sourceList.begin(), m_sourceList.end(),
+                [name, includeTombstone](const SourceDetailsInternal& sd)
+                {
+                    return Utility::ICUCaseInsensitiveEquals(sd.Name, name) &&
+                        (!sd.IsTombstone || includeTombstone);
+                });
+        }
+
+        bool SourceListInternal::ContainsCurrentSource(std::string_view name)
+        {
+            return FindSource(name) != m_sourceList.end();
+        }
+
+        SourceDetailsInternal& SourceListInternal::GetCurrentSource(std::string_view name)
+        {
+            auto itr = FindSource(name);
+            THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), itr == m_sourceList.end());
+            return *itr;
+        }
+
+        bool SourceListInternal::ContainsSource(std::string_view name)
+        {
+            return FindSource(name, true) != m_sourceList.end();
+        }
+
+        SourceDetailsInternal& SourceListInternal::GetSource(std::string_view name)
+        {
+            auto itr = FindSource(name, true);
+            THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), itr == m_sourceList.end());
+            return *itr;
+        }
+
+        void SourceListInternal::AddSource(const SourceDetailsInternal& details)

```

Comment:
>AddSource [](start = 33, length = 9)

This function will need to remove any existing tombstone before adding the details, or they would be blocked behind said tombstone.